### DISCUSSION TITLE
Force filesystem sync to ensure that files are actually written on disk

### DISCUSSION THREAD

**Discussion by polodroid74** [AxisCommunications/discussions#192] [ACAP] [answered]
Hi,
Is there a way to ensure that a file which is opened, modified and closed by an ACAP application is effectively written on disk even if a power outage occurs.

Indeed, I run an ACAP application written in C++ using std::ofstream class to write file on disk. Here is an example : 

```
std::ofstream stream;
stream.open();
//Do some stuff on the stream
stream.close();
```

However, in case of power outage, the previous example does not warranty that the file is effectively written on disk.
I know this is common "issue" on UNIX-like systems ([https://man7.org/linux/man-pages/man2/fsync.2.html](https://man7.org/linux/man-pages/man2/fsync.2.html)) but I was wondering if, on AXIS devices, there was a way to force the file to be written as soon as we close the stream. 
Can someone provide any help or tips on this ?
NB : My application is developped using ACAP 3.5 on an AXIS P-1378-LE camera (AXIS OS version 10.12.135)

Regards,
Paul

**Comment by daniel-falk** [comment, marked as answer]
@polodroid74 There are two aspects to it: 1) Make sure the file is never partially written, 2) verify that the file is actually written. I don't think there is an Axis-specific way of doing this, and for 2) I don't think there is any way of doing it at all (if it is really important you should probably talk to the people creating engines for databases, they are struggling with that very specific issue).

Different cameras has different filesystems, I know of at least cameras with `UBIFS` and cameras with `ext4`. Regarding point 2, different cameras might have different hardware that deals differently with buffering in hardware registers, I don't know, only guessing here...

The standard solution on POSIX systems to achieve atomicity is to write to a new file, call `fsync()` on the new file and then rename the new file to overwrite the old file (since a rename is atomic). This does not guarantee that the file is written, but it guarantees the the state of the file is either the old state or the new, not something in between which could corrupt the files or cause a truncated file.

All Axis cameras include `glib`, so we are usually using the `g_file_set_contents()` function in `glib` which use this method. (Note that you can also use `g_file_set_contents_full()` with the  `G_FILE_SET_CONTENTS_CONSISTENT` flag set, but this was added in version 2.66 and the latest firmware for e.g. companion line has version 2.60 of glib).

This should as I understand it result in atomicity on both [ext4](https://blogs.gnome.org/alexl/2009/03/16/ext4-vs-fsync-my-take/) and [UBIFS](http://www.linux-mtd.infradead.org/doc/ubifs.html#L_writeback).

Further you can call `fsync` on the parent directory which increases the chances that the file path actually points to the new file in case of a power loss. But as I understand it, none of these file systems (even `ext4` that is journald) guarantees that the data is actually written to the persistent flash.

**Reply by polodroid74** [reply]
Thanks a lot for all these detailed information. I'll try to fix my issue using your recommendations.
